sklearn/feature_extraction/_hashing_fast.pyx

```diff
@@ -35,7 +35,7 @@ def transform(raw_X, Py_ssize_t n_features, dtype,
     # and values arrays ourselves. Use a Py_ssize_t capacity for safety.
     cdef Py_ssize_t capacity = 8192     # arbitrary
     cdef cnp.int64_t size = 0
-    cdef cnp.ndarray values = np.empty(capacity, dtype=dtype)
+    cdef double[:] values = np.empty(capacity, dtype=np.float64)
```

Comment:
We likely need to keep this as `cnp.ndarray` because of how the `dtype` argument is passed in.

With this PR, `_hashing_transform` would always return a `float64`. Later, when the `csr_matrix` is formed, it will make a copy if `self.dtype=np.float32`:

https://github.com/scikit-learn/scikit-learn/blob/4f85597ac5216ea21f1a8b343f6487bd3c9eb27e/sklearn/feature_extraction/_hash.py#L185-L189

With the current code in `main` and `dtype=np.float32`, the `values` would be `float32` and `csr_matrix` does not need to make a copy.

Reply:
I see. That means that dtype can only be float32 or float64? In one of the tests I observed that the dtype passed to the fast hash transform function was a string.

Reply:
Since we declare value as a double I think a fused type like floating would also not be totally suitable.

Reply:
I think the dtype can be anything. `float32`, `float64`, `int32`, etc. I do not think it's worth fusing everything because we can missing a dtype someone might be using.

It's okay for `value` to be double. When setting `value[size] = value`, it would do C-style casting.

Reply:
Sounds good. I am going to close this PR as no change needs to be made here.